### PR TITLE
Refactors JDBC driver to parse TLS options in VitessConnection class

### PR DESCRIPTION
**Pull request by steve-perkins**
This is a follow-up to PR #2349, including some refactoring cleanup that was omitted from the original PR because that one had grown excessively large and complex.

In the JDBC driver, responsibility for parsing SSL/TLS-related optional settings now resides in the `VitessConnection` class.  As part of making that work, the gRPC setup is now triggered by `VitessConnection.connect()` rather than the constructor method.

@bbeaudreault is the most familiar with the prior Slack discussion leading up to this.

---
This change is [Reviewable](https://reviewable.io/reviews/youtube/vitess/2529)

### CODE REVIEW

**Comment by steve-perkins** [comment]
There's some ongoing discussion in Slack, but I would recommend that we increment the version number of this JDBC driver artifact prior to merging this pull request.  

It's currently sitting at `1.0-SNAPSHOT`, which has been the version number since the code was initially donated from Flipkart.  Even an increment to `1.1.0-SNAPSHOT` would be nice, since:

1. It would bring the convention in line with semver standards, and
2. There's no easy way to tell which version of the driver artifact you have in hand (and there have been a LOT of changes since the initial Flipkark contribution).  So some new version bump would at least give users a milepost marker to gauge against.

**Comment by steve-perkins** [comment]
Based on discussion with @harshit-gangal in Slack, I've incremented the version number of the Java client artifacts from `1.0-SNAPSHOT` to `1.1.0-SNAPSHOT`.  Discussion of a longer-term versioning strategy is ongoing, and outside the scope of this PR.

Because of the way that these Maven build files are structured, bumping the version number requires a search-and-replace that touches multiple files.  That's a bit unfortunate (probably part of the reason why no one's wanted to increase the version number before!), and this could be centralized in a single place with some minor Maven restructuring.  However, that work should probably wait until after more decisions are made regarding the long-term versioning strategy.

**Comment by harshit-gangal** [comment]
Still, the version number should be same for all the artifacts. We should
not differ on that.

**Comment by harshit-gangal** [comment]
Reviewed 17 of 17 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessConnection.java, line 82 at r1](https://reviewable.io:443/reviews/youtube/vitess/2529#-KcST-7cIxG65_CHuLFr:-KcST-7cIxG65_CHuLFs:b-ks5uka) ([raw file](https://github.com/youtube/vitess/blob/34a35fc2baec8a440a64a4445fd0508670167044/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessConnection.java#L82)):*
> ```Java
> 
>     public void connect() {
>         this.vTGateConnections = new VitessVTGateManager.VTGateConnections(this);
> ```

What benefit you get to move it to separate method?

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessDatabaseMetaData.java, line 29 at r1](https://reviewable.io:443/reviews/youtube/vitess/2529#-KcSSiZ4ysjMnph9z-Yr:-KcSSiZ4ysjMnph9z-Ys:b132jzy) ([raw file](https://github.com/youtube/vitess/blob/34a35fc2baec8a440a64a4445fd0508670167044/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessDatabaseMetaData.java#L29)):*
> ```Java
> 
>     public String getURL() throws SQLException {
>         return this.connection.getUrl().getUrl();
> ```

Check for VitessJDBCUrl to be NotNull

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2529)*

**Comment by bbeaudreault** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessConnection.java, line 82 at r1](https://reviewable.io:443/reviews/youtube/vitess/2529#-KcST-7cIxG65_CHuLFr:-KcSZ-jQNAkoDutPQOoG:b-goqvan) ([raw file](https://github.com/youtube/vitess/blob/34a35fc2baec8a440a64a4445fd0508670167044/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessConnection.java#L82)):*
<details><summary><i>Previously, harshit-gangal wrote…</i></summary><blockquote>

What benefit you get to move it to separate method?
</blockquote></details>

We thought it a bad practice to pass (this) in to another constructor while the object itself was still constructing. I think it should work in our case currently, but may not always be true in the future.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessJDBCUrl.java, line 163 at r1](https://reviewable.io:443/reviews/youtube/vitess/2529#-KcS_6_cMjTRFUAiB1sX:-KcS_6_cMjTRFUAiB1sY:bggiekn) ([raw file](https://github.com/youtube/vitess/blob/34a35fc2baec8a440a64a4445fd0508670167044/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessJDBCUrl.java#L163)):*
> ```Java
>         this.trustStore = caseInsensitiveKeyLookup(info, Constants.Property.TRUSTSTORE);
>         this.trustStorePassword = caseInsensitiveKeyLookup(info, Constants.Property.TRUSTSTORE_PASSWORD);
>         this.trustAlias = caseInsensitiveKeyLookup(info, Constants.Property.TRUST_ALIAS);
> ```

I think all of these can be removed now, right? Including the getters you added below as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2529)*

**Comment by harshit-gangal** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessConnection.java, line 82 at r1](https://reviewable.io:443/reviews/youtube/vitess/2529#-KcST-7cIxG65_CHuLFr:-KcSu9Dd6ynKimgbBhVS:b2ni9bh) ([raw file](https://github.com/youtube/vitess/blob/34a35fc2baec8a440a64a4445fd0508670167044/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessConnection.java#L82)):*
<details><summary><i>Previously, bbeaudreault (Bryan Beaudreault) wrote…</i></summary><blockquote>

We thought it a bad practice to pass (this) in to another constructor while the object itself was still constructing. I think it should work in our case currently, but may not always be true in the future.
</blockquote></details>

ok.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessJDBCUrl.java, line 163 at r1](https://reviewable.io:443/reviews/youtube/vitess/2529#-KcS_6_cMjTRFUAiB1sX:-KcSumjeIfeCS1CyvXw7:b478c64) ([raw file](https://github.com/youtube/vitess/blob/34a35fc2baec8a440a64a4445fd0508670167044/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessJDBCUrl.java#L163)):*
<details><summary><i>Previously, bbeaudreault (Bryan Beaudreault) wrote…</i></summary><blockquote>

I think all of these can be removed now, right? Including the getters you added below as well.
</blockquote></details>

This the only place where the URL is parsed. If I am right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2529)*

**Comment by bbeaudreault** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessJDBCUrl.java, line 163 at r1](https://reviewable.io:443/reviews/youtube/vitess/2529#-KcS_6_cMjTRFUAiB1sX:-KcSwK9NGlBXITEwU5FG:bkc1dua) ([raw file](https://github.com/youtube/vitess/blob/34a35fc2baec8a440a64a4445fd0508670167044/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessJDBCUrl.java#L163)):*
<details><summary><i>Previously, harshit-gangal wrote…</i></summary><blockquote>

This the only place where the URL is parsed. If I am right?
</blockquote></details>

The URL is parsed at https://github.com/youtube/vitess/blob/34a35fc2baec8a440a64a4445fd0508670167044/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessJDBCUrl.java#L260 and all values are placed into the Properties object.

The Properties object is returned and used in ConnectionProperties. Accessing of properties should happen in ConnectionProperties.

Basically the goal here was to make VitessJDBCUrl be used for generic parsing of URL parameters into Properties key/values, merging with existing properties where appropriate. I left the host/user/password in there because they were part of the url structure (not parameters). But they also could be moved if we wanted.

All reading of configuration parameters should happen in ConnectionProperties. This provides a centralized place where you can see all the properties, their descriptions, their default values, and whether they are required.

@steve-perkins did part of that, moving the getters into ConnectionProperties. But he should also remove them from here since now they are redundant.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2529)*

**Comment by steve-perkins** [comment]
The group discussed it Slack yesterday (2017-03-02, 2:00 PM EST), and the consensus was to bump it to `1.1.0-SNAPSHOT` for simplicity right now.  Longer term, a strategy is desired to align the JDBC driver version with the overall Vitess version number, and move to release rather than snapshot versions.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2529#-:-KeJqDK0v90t32LNmzOK:b-gx469a)*

**Comment by steve-perkins** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessDatabaseMetaData.java, line 29 at r1](https://reviewable.io:443/reviews/youtube/vitess/2529#-KcSSiZ4ysjMnph9z-Yr:-KeKTsvqBI_8TJE6unt9:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/34a35fc2baec8a440a64a4445fd0508670167044/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessDatabaseMetaData.java#L29)):*
<details><summary><i>Previously, harshit-gangal wrote…</i></summary><blockquote>

Check for VitessJDBCUrl to be NotNull
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2529)*

**Comment by bbeaudreault** [comment]
Reviewed 17 of 17 files at r1, 5 of 5 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[java/example/pom.xml, line 17 at r1](https://reviewable.io:443/reviews/youtube/vitess/2529#-KeKfblmk3DotV0YCy5X:-KeKfblmk3DotV0YCy5Y:bbdkf3p) ([raw file](https://github.com/youtube/vitess/blob/34a35fc2baec8a440a64a4445fd0508670167044/java/example/pom.xml#L17)):*
> ```XML
>       <groupId>com.flipkart.vitess</groupId>
>       <artifactId>vitess-connector-java</artifactId>
>       <version>1.1.0-SNAPSHOT</version>
> ```

I'm not going to insist on this for now -- @michael-berlin can handle it when he does the eventual real versioning -- but I think it would be best to make version a property:

```
<properties>
  <vitess.version>1.1.0-SNAPSHOT<vitess.version>
</properties>
```

Then in all of these poms you can just make it look like this:

```
<dependency>
      <groupId>com.youtube.vitess</groupId>
      <artifactId>client</artifactId>
      <version>${vitess.version}</version>
 </dependency>
```

Will result in less changes in the future.  The one place I don't think this works is in the <parent> block. But also we don't *really* need to bump the parent version unless we change the packaging, so not a huge issue. We can just bump all the child versions when we want to.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/ConnectionProperties.java, line 52 at r1](https://reviewable.io:443/reviews/youtube/vitess/2529#-KeKgU0zcvGwzwWqvx4-:-KeKgU0zcvGwzwWqvx40:b-2n374p) ([raw file](https://github.com/youtube/vitess/blob/34a35fc2baec8a440a64a4445fd0508670167044/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/ConnectionProperties.java#L52)):*
> ```Java
> 
>     // TLS-related configs
>     private BooleanConnectionProperty useSSL = new BooleanConnectionProperty(
> ```

This is more of just a question, potentially out of scope for this PR:

Recently I was looking at the basic (non-vitess) JDBC SSL configs and saw the `verifyServerCertificate` option. Should we be supporting that here? It seems like a useful option for situations where you want encryption but trust your network.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2529)*

**Comment by bbeaudreault** [comment]
:lgtm:

I added 1 optional comment and had a question for you. Thanks for the cleanup!

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2529#-:-KeKjOysUS9JfTJEwP6U:b-ev91e5)*

**Comment by steve-perkins** [comment]
Review status: 17 of 18 files reviewed at latest revision, 5 unresolved discussions.

---

*[java/example/pom.xml, line 17 at r1](https://reviewable.io:443/reviews/youtube/vitess/2529#-KeKfblmk3DotV0YCy5X:-KeKobA1jbm5uiXKk2SF:b-e5gtol) ([raw file](https://github.com/youtube/vitess/blob/34a35fc2baec8a440a64a4445fd0508670167044/java/example/pom.xml#L17)):*
<details><summary><i>Previously, bbeaudreault (Bryan Beaudreault) wrote…</i></summary><blockquote>

I'm not going to insist on this for now -- @michael-berlin can handle it when he does the eventual real versioning -- but I think it would be best to make version a property:

```
<properties>
  <vitess.version>1.1.0-SNAPSHOT<vitess.version>
</properties>
```

Then in all of these poms you can just make it look like this:

```
<dependency>
      <groupId>com.youtube.vitess</groupId>
      <artifactId>client</artifactId>
      <version>${vitess.version}</version>
 </dependency>
```

Will result in less changes in the future.  The one place I don't think this works is in the <parent> block. But also we don't *really* need to bump the parent version unless we change the packaging, so not a huge issue. We can just bump all the child versions when we want to.
</blockquote></details>

Agreed (although yeah, I would prefer that this be part of the ultimate "real versioning" effort).

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/ConnectionProperties.java, line 52 at r1](https://reviewable.io:443/reviews/youtube/vitess/2529#-KeKgU0zcvGwzwWqvx4-:-KeKomk1cYnyyOJ4EVy5:b-83w6pa) ([raw file](https://github.com/youtube/vitess/blob/34a35fc2baec8a440a64a4445fd0508670167044/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/ConnectionProperties.java#L52)):*
<details><summary><i>Previously, bbeaudreault (Bryan Beaudreault) wrote…</i></summary><blockquote>

This is more of just a question, potentially out of scope for this PR:

Recently I was looking at the basic (non-vitess) JDBC SSL configs and saw the `verifyServerCertificate` option. Should we be supporting that here? It seems like a useful option for situations where you want encryption but trust your network.
</blockquote></details>

Sure.  I've never used this option personally, but I see no reason for Vitess not to support it.  As long as this particular pull request has lingered around though, I'd rather see it done is a separate PR.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessJDBCUrl.java, line 163 at r1](https://reviewable.io:443/reviews/youtube/vitess/2529#-KcS_6_cMjTRFUAiB1sX:-KeKp8XQc0o4uZFYySpi:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/34a35fc2baec8a440a64a4445fd0508670167044/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessJDBCUrl.java#L163)):*
<details><summary><i>Previously, bbeaudreault (Bryan Beaudreault) wrote…</i></summary><blockquote>

The URL is parsed at https://github.com/youtube/vitess/blob/34a35fc2baec8a440a64a4445fd0508670167044/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessJDBCUrl.java#L260 and all values are placed into the Properties object.

The Properties object is returned and used in ConnectionProperties. Accessing of properties should happen in ConnectionProperties.

Basically the goal here was to make VitessJDBCUrl be used for generic parsing of URL parameters into Properties key/values, merging with existing properties where appropriate. I left the host/user/password in there because they were part of the url structure (not parameters). But they also could be moved if we wanted.

All reading of configuration parameters should happen in ConnectionProperties. This provides a centralized place where you can see all the properties, their descriptions, their default values, and whether they are required.

@steve-perkins did part of that, moving the getters into ConnectionProperties. But he should also remove them from here since now they are redundant.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2529)*

**Comment by bbeaudreault** [comment]
LGTM.

@michael-berlin I'm going to be away when  you do the `io.vitess` refactor. Can you also at that time do the change I suggested above? That is, use a property for the version instead of hardcoding it in every pom multiple times

@steve-perkins agreed, `verifyServerCertificate` can be a separate PR just wanted to see if there was a reason.

**Comment by michael-berlin** [comment]
> @michael-berlin I'm going to be away when you do the io.vitess refactor. Can you also at that time do the change I suggested above? That is, use a property for the version instead of hardcoding it in every pom multiple times

Ack.

**Comment by steve-perkins** [comment]
Okay, all comments have been addressed, git commits squashed, and Travis CI is green.  No further changes are planned on my end unless anyone has anything further.

---

Review status: 17 of 18 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2529#-:-KeL4Crmm0sN8mqZqk3E:b-oor9ng)*

**Comment by michael-berlin** [comment]
LGTM since @bbeaudreault signed off on the change.

Merging now.
